File: packages/javascript/lib/word/wordLengthRange.ts
import { clamp } from '../_internal/utils.js';
import { RAND_LENGTH_MAX, RAND_LENGTH_MIN } from '../constants.js';
import type { WordLanguageOption, WordThemeOption } from '../_types/global.js';
import { WORD_LANGUAGES } from './data/index.js';
import { naturalRange } from './wordGenerator.js';

/**
 * Shortest and longest word the language's pools hold, in characters, which is
 * what `randWord` falls back to when `minLength` or `maxLength` is omitted.
 * Narrowing the theme narrows the range, because a theme is a pool of its own.
 *
 * @example
 * wordLengthRange('ko'); // [1, 5]
 * wordLengthRange('ko', 'animal'); // [2, 4]
 * wordLengthRange('en'); // [2, 12]
 */
export function wordLengthRange(
	language: WordLanguageOption = 'all',
	theme: WordThemeOption = 'all'
): [number, number] {
	const languages = language === 'all' ? WORD_LANGUAGES : [language];
	let min = Infinity;
	let max = 0;

	for (const code of languages) {
		const [low, high] = naturalRange(code, theme);

		min = Math.min(min, low);
		max = Math.max(max, high);
	}

	return [
		clamp(min, RAND_LENGTH_MIN, RAND_LENGTH_MAX),
		clamp(max, RAND_LENGTH_MIN, RAND_LENGTH_MAX)
	];
}

File: packages/dart/lib/src/word/word_length_range.dart
import 'package:randino/src/constants.dart';
import 'package:randino/src/internal/utils.dart';
import 'package:randino/src/types.dart';
import 'package:randino/src/word/data/index.dart';
import 'package:randino/src/word/word_generator.dart';

/// Shortest and longest word the language's pools hold, in characters.
///
/// This is what `randWord` falls back to when `minLength` or `maxLength` is
/// omitted. Narrowing the [theme] narrows the range, because a theme is a pool
/// of its own. A null [language] or [theme] spans every one of them.
///
/// ```dart
/// wordLengthRange(language: WordLanguage.ko); // LengthRange(1, 4)
/// wordLengthRange(language: WordLanguage.ko, theme: WordTheme.food); // LengthRange(1, 4)
/// wordLengthRange(language: WordLanguage.en); // LengthRange(3, 11)
/// ```
LengthRange wordLengthRange({WordLanguage? language, WordTheme? theme}) {
  final languages = language == null ? wordLanguages : <WordLanguage>[language];
  var min = 1 << 30;
  var max = 0;

  for (final code in languages) {
    final range = naturalRange(code, theme);

    if (range.min < min) min = range.min;
    if (range.max > max) max = range.max;
  }

  return LengthRange(
    clampInt(min, randLengthMin, randLengthMax),
    clampInt(max, randLengthMin, randLengthMax),
  );
}
